Captured video frames are leaked if Web process media player is destroyed before message is received

kkinnunen-apple opened the pull request:
#### 3da3dca6459daf055e08d25a0ee8a779037b6b26
<pre>
Captured video frames are leaked if Web process media player is destroyed before message is received
<a href="https://bugs.webkit.org/show_bug.cgi?id=279671">https://bugs.webkit.org/show_bug.cgi?id=279671</a>
<a href="https://rdar.apple.com/problem/135948989">rdar://problem/135948989</a>

Reviewed by NOBODY (OOPS!).

GPU process would put the captured VideoFrame to the video frame heap
and send the reference to the WebContent process media player. If
the MediaPlayerPrivateRemote was destroyed while the message was in transit,
the message would be ignored and the video frame would leak in the
heap.

Make the GPUP -&gt; WCP video frame offer messages asynchronous with
replies. Since asynchronous message replies are always delivered, this
fixes the memory leak. Remove the video frame from the heap if
the frame was not adopted.

This concerns following video frame offer messages:
    Messages::MediaPlayerPriveteRemote::PushVideoFrameMetadata
    Messages::LibWebRTCCodecs::CompletedDecoding
    Messages::RemoteCaptureSampleManager::videoFrameAvailable

* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
(WebKit::RemoteVideoFrameObjectHeap::addWithCompletion):
(WebKit::RemoteVideoFrameObjectHeap::add):
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.h:
* Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerProxyCocoa.mm:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerOnNewVideoFrameMetadata):
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
(WebKit::LibWebRTCCodecsProxy::createDecoderCallback):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::videoFrameForCurrentTime):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.messages.in:
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
(WebKit::RemoteVideoFrameProxy::properties):
(WebKit::RemoteVideoFrameProxy::~RemoteVideoFrameProxy):
(WebKit::releaseRemoteVideoFrameProxy): Deleted.
(WebKit::RemoteVideoFrameProxy::releaseUnused): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxyProperties.h:
* Source/WebKit/WebProcess/GPU/media/cocoa/MediaPlayerPrivateRemoteCocoa.mm:
(WebKit::MediaPlayerPrivateRemote::pushVideoFrameMetadata):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::LibWebRTCCodecs::completedDecoding):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.h:
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.messages.in:
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.cpp:
(WebKit::RemoteCaptureSampleManager::videoFrameAvailable):
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.h:
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.messages.in:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3da3dca6459daf055e08d25a0ee8a779037b6b26

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66855 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46230 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19477 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70890 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17988 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54029 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17767 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53632 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12104 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69922 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42534 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57843 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34177 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39205 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16342 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61119 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15576 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72591 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10812 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14927 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61034 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10844 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57900 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61191 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8877 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2501 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42037 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43114 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44297 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42857 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.h:
```diff
@@ -50,7 +50,13 @@ class RemoteVideoFrameObjectHeap final : public IPC::WorkQueueMessageReceiver {
-    RemoteVideoFrameProxy::Properties add(Ref<WebCore::VideoFrame>&&);
+    // Returns completion that can be passed to asynchronous create messages. Upon successful
+    // reply of the message, the frame will be added. Upon cancellation of the message reply,
+    // the frame will be discarded.
```
I wonder whether this could create potential races between the IPC async reply (that will add frames to the map) and a message M from WebProcess that would make use of the ID right away.
Couldn't the message M be processed before the async reply message and get no frame from its ID?

kkinnunen-apple replied:
That's why the "object heap" structure is being used.
The message M contains "read reference" to the video frame.
The reader will wait in RemoteVideFrameObjectHeap::get Until the read reference is available.
When the async message reply completes, the resource gets added to the object heap and the read reference will be available.

youennf commented on Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
```diff
@@ -495,18 +495,15 @@ void LibWebRTCCodecs::failedDecoding(VideoDecoderIdentifier decoderIdentifier)
-    // Adopt RemoteVideoFrameProxy::Properties to RemoteVideoFrameProxy instance before the early outs, so that the reference gets adopted.
-    // Typically RemoteVideoFrameProxy::Properties&& sent to destinations that are already removed need to be handled separately.
-    // LibWebRTCCodecs is not ever removed, so we do not do this. However, if it ever is, LibWebRTCCodecs::handleMessageToRemovedDestination()
-    // needs to be implemented.
```
I like the idea of removing the risk for memory leaks like this.
I am wondering though whether it would not be better to keep IDs being generated where the frames are created (GPUProcess) and have the completion handler remove the frame from the heap in case something wrong is happening.

kkinnunen-apple replied:
That's the opposite of what is the target of this patch.
If we generate the IDs in GPUP, we optimize for the slow path: the case where GPUP creates the id and adds it to the heap.
We need to optimize for the fast path: WCP creates the id and sends the message that will snapshot the video frame from WebGL context or from media player. This way WCP can continue using the video frame + id before that happens in the GPUP.

youennf commented on Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.h:
```diff
@@ -50,7 +50,13 @@ class RemoteVideoFrameObjectHeap final : public IPC::WorkQueueMessageReceiver {
+    CompletionHandler<void(RemoteVideoFrameIdentifier)> addWithCompletion(Ref<WebCore::VideoFrame>&&);
+    // Adds video frame with a known reference.
+    // returns true on success, false if the reference existed already.
+    bool add(RemoteVideoFrameReference&&, Ref<WebCore::VideoFrame>&&);
```
With these two add methods, we could potentially have IDs generated by WebProcess and GPUProcess.
I guess the plan is for WebProcess to be the only one creating remote frame IDs?

I tend to prefer being able to create IDs where frames are created (GPU process mostly, could be useful in WebProcess as well).
This ensures that when you have an ID, the GPUProcess has the frame.
It seems more complex to have WebProcess generating IDs, and potentially using them without waiting for GPUProcess to actually register the IDs with their frames.

If we want to have no collision and be able to generate IDs in both processes, we could add a specific mechanism (even IDs are generated by WebProcess, odd IDs are generated by GPUProcess).

kkinnunen-apple replied:
> If we want to have no collision and be able to generate IDs in both processes, we could add a specific mechanism (even IDs are generated by WebProcess, odd IDs are generated by GPUProcess).

We can dedicate id prefix to GPU and id prefix to WCP, but it's a bit orthogonal to this patch.
It doesn't buy us much anything other than the inverse case, e.g. the fact that we'd remove the unused entry, vs. we'd add the used entry.

youennf commented on Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.mm:
```diff
@@ -35,7 +35,10 @@
 RefPtr<WebCore::VideoFrame> RemoteVideoFrameObjectHeap::get(RemoteVideoFrameReadReference&& read)
 {
-    return m_heap.read(WTFMove(read), 0_s);
+    RefPtr result = m_heap.read(WTFMove(read), 5_s);
```
This is probably the change that fixes the potential race, correct?
This can introduce potential small locks between the camera output thread and the local rendering thread.
Currently, these two are mostly independent with each other.

kkinnunen-apple replied:
Yes and no.

It's just that if, say, RemoteGraphicsContextGL connection inside its work queue creates the video frame, it won't be available in WebRTC encoding thread until the RemoteGraphicsContextGL is done generating the video frame.

Same with media player: media player frames are snapshot in media player work queue (e.g. GPUP main thread), but they are used in RemoteRenderingBackend thread. So RRB needs to wait.

Currently the wait has been in WCP, in form of sync call. Later, that sync is removed.

youennf commented on Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
```diff
@@ -133,12 +133,13 @@
-            connection->send(Messages::LibWebRTCCodecs::CompletedDecoding { identifier, timeStamp, timeStampNs, WTFMove(properties) }, 0);
+            auto properties = RemoteVideoFrameProxy::properties(newFrame);
+            auto [newFrameIdentifier, completion] = videoFrameObjectHeap->addWithCompletion(WTFMove(newFrame));
+            connection->sendWithAsyncReply(Messages::LibWebRTCCodecs::CompletedDecoding { identifier, timeStamp, timeStampNs, { newFrameIdentifier, WTFMove(properties) } }, WTFMove(completion), 0);
```
I guess these changes are not strictly needed anymore for your web process changes.

That said, this still ties cameras/decoders thread with rendering threads for video frames.
Why not simply adding the video frame synchronously here and have the completion handler returned by the object heap to remove the frame if called with false (aka WebProcess adoption failed)?
Or split these changes in its own PR.

kkinnunen-apple replied:
> Why not simply adding the video frame synchronously here

That's what it does now.

youennf commented on Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
```diff
@@ -91,13 +91,31 @@ void RemoteVideoFrameObjectHeap::close()
-    return properties;
+    // High bit of the identifiers are reserved for identifiers allocated in the object holder namespace, e.g. GPU process.
+    static std::atomic<uint64_t> rawIdentifier;
+    RemoteVideoFrameIdentifier identifier { ++rawIdentifier | (static_cast<uint64_t>(1) << 63) };
```
Maybe we should add a GPUProcess generate method  and a WebProcess generate method to make the code code clearer.

youennf commented on Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
```diff
@@ -952,8 +952,12 @@ void RemoteMediaPlayerProxy::videoFrameForCurrentTimeIfChanged(CompletionHandler
-            result = m_videoFrameObjectHeap->add(videoFrame.releaseNonNull());
+        if (videoFrame) {
+            result = RemoteVideoFrameProxy::properties(*videoFrame);
+            bool success = m_videoFrameObjectHeap->add({ newFrame, 0 }, videoFrame.releaseNonNull());
```
Should we make the second parameter have a default value of 0?

youennf commented on Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.h:
```diff
@@ -50,7 +50,22 @@ class RemoteVideoFrameObjectHeap final : public IPC::WorkQueueMessageReceiver {
+
+    struct AddWithCompletionResult {
+        RemoteVideoFrameIdentifier identifier;
+        CompletionHandler<void(bool)> completion;
```
```suggestion
        CompletionHandler<void(bool willRemoveFrame)> completion;
```

kkinnunen-apple committed:
Captured video frames are leaked if Web process media player is destroyed before message is received
https://bugs.webkit.org/show_bug.cgi?id=279671
rdar://problem/135948989

Reviewed by NOBODY (OOPS!).

GPU process would put the captured VideoFrame to the video frame heap
and send the reference to the WebContent process media player. If
the MediaPlayerPrivateRemote was destroyed while the message was in transit,
the message would be ignored and the video frame would leak in the
heap.

Make the GPUP -> WCP video frame offer messages asynchronous with
replies. Since asynchronous message replies are always delivered, this
fixes the memory leak. Remove the video frame from the heap if
the frame was not adopted.

This concerns following video frame offer messages:
    Messages::MediaPlayerPriveteRemote::PushVideoFrameMetadata
    Messages::LibWebRTCCodecs::CompletedDecoding
    Messages::RemoteCaptureSampleManager::videoFrameAvailable

* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
(WebKit::RemoteVideoFrameObjectHeap::addWithCompletion):
(WebKit::RemoteVideoFrameObjectHeap::add):
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.h:
* Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerProxyCocoa.mm:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerOnNewVideoFrameMetadata):
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
(WebKit::LibWebRTCCodecsProxy::createDecoderCallback):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::videoFrameForCurrentTime):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.messages.in:
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
(WebKit::RemoteVideoFrameProxy::properties):
(WebKit::RemoteVideoFrameProxy::~RemoteVideoFrameProxy):
(WebKit::releaseRemoteVideoFrameProxy): Deleted.
(WebKit::RemoteVideoFrameProxy::releaseUnused): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxyProperties.h:
* Source/WebKit/WebProcess/GPU/media/cocoa/MediaPlayerPrivateRemoteCocoa.mm:
(WebKit::MediaPlayerPrivateRemote::pushVideoFrameMetadata):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::LibWebRTCCodecs::completedDecoding):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.h:
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.messages.in:
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.cpp:
(WebKit::RemoteCaptureSampleManager::videoFrameAvailable):
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.h:
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.messages.in:
